File: Leetcode/3Sum/solution.cpp
#include <iostream>
#include <string>
#include <vector>
#include <set>
#include <algorithm>
using namespace std;

class Solution { 
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        int i, j, k, sum;
        vector<vector<int>> ans;
        sort(nums.begin(), nums.end());

        for (i=0; i<nums.size(); i++) {
            if (i>0 and nums[i]==nums[i-1]) continue;
            j = i+1;
            k = nums.size()-1;
            while (j < k) {
                sum = nums[i] + nums[j] + nums[k];
                if (sum == 0) {
                    ans.push_back({nums[i], nums[j], nums[k]});
                    while (j < k and nums[j+1] == nums[j]) {
                        j++;
                    }
                    while (k > j and nums[k-1] == nums[k]) {
                        k--;
                    }
                    j++;
                    k--;
                } else if (sum > 0) {
                    k--;
                } else { // sum < 0
                    j++;
                }
            }
        }
        
        return ans;
    }
};

int main() {
    Solution solution;
    vector<int> nums{-1,0,1,2,-1,-4};

    vector<vector<int>> ans = solution.threeSum(nums);
    for (int i=0; i<ans.size(); i++) {
        for (int j=0; j<ans[i].size(); j++) {
            cout<<ans[i][j]<<" ";
        }
        cout<<endl;
    }
    return 0;
}

File: Leetcode/3Sum/java/src/solution.java
import java.util.*;

class Solution {
    private List<List<Integer>> twoSum(int[] nums, int target, List<List<Integer>> ans) {
        int left=0, right=nums.length-1;

        while (left < right) {
            int cur = nums[left] + nums[right];
            if (cur == target) {
                ans.add(Arrays.asList(-1*target, nums[left], nums[right]));

                left++;
                while(left<right && nums[left]==nums[left-1]) left++;
                right--;
                while(left<right && nums[right]==nums[right+1]) right--;
            } else if (cur < target) {
                left++;
            } else { // cur > target
                right--;
            }
        }

        return ans;
    }
    public List<List<Integer>> threeSum(int[] nums) {
        Arrays.sort(nums);
        List<List<Integer>> ans = new ArrayList<>();

        for (int i=0; i<nums.length; i++) {
            if (i>0 && nums[i]==nums[i-1]) continue;
            
            List<List<Integer>> res = twoSum(Arrays.copyOfRange(nums, i+1, nums.length), nums[i] * -1, ans);
        }
        
        return ans;
    }
}

class Main {
    static public void main(String[] args) {
        Solution solution = new Solution();
        int[] nums = {0,-1,0,1,2,-1,1,-4};

        List<List<Integer>> ans = solution.threeSum(nums);
    }
}